pkg/network/http/ebpf_gotls.go

```diff
@@ -0,0 +1,212 @@
+
+	writeFuncName      = "uprobe__crypto_tls_Conn_Write"
+	readFuncName       = "uprobe__crypto_tls_Conn_Read"
+	readReturnFuncName = "uprobe__crypto_tls_Conn_Read__return"
+	closeFuncName      = "uprobe__crypto_tls_Conn_Close"
+
+	writeProbe      = "uprobe/crypto/tls.(*Conn).Write"
+	readProbe       = "uprobe/crypto/tls.(*Conn).Read"
+	readReturnProbe = "uprobe/crypto/tls.(*Conn).Read/return"
+	closeProbe      = "uprobe/crypto/tls.(*Conn).Close"
+)
+
+var functionsConfig = map[string]bininspect.FunctionConfiguration{
+	bininspect.WriteGoTLSFunc: {
+		IncludeReturnLocations: false,
+		ParamLookupFunction:    lookup.GetWriteParams,
+	},
+	bininspect.ReadGoTLSFunc: {
+		IncludeReturnLocations: true,
+		ParamLookupFunction:    lookup.GetReadParams,
+	},
+	bininspect.CloseGoTLSFunc: {
+		IncludeReturnLocations: false,
+		ParamLookupFunction:    lookup.GetCloseParams,
+	},
+}
+
+var structFieldsLookupFunctions = map[bininspect.FieldIdentifier]bininspect.StructLookupFunction{
+	bininspect.StructOffsetTLSConn:     lookup.GetTLSConnInnerConnOffset,
+	bininspect.StructOffsetTCPConn:     lookup.GetTCPConnInnerConnOffset,
+	bininspect.StructOffsetNetConnFd:   lookup.GetConnFDOffset,
+	bininspect.StructOffsetNetFdPfd:    lookup.GetNetFD_PFDOffset,
+	bininspect.StructOffsetPollFdSysfd: lookup.GetFD_SysfdOffset,
+}
+
+type GoTLSProgram struct {
+	manager *manager.Manager
+}
+
+// Static evaluation to make sure we are not breaking the interface.
+var _ subprogram = &GoTLSProgram{}
+
+func NewGoTLSProgram(c *config.Config) (*GoTLSProgram, error) {
+	if !c.EnableHTTPSMonitoring {
+		return nil, nil
+	}
+	return &GoTLSProgram{}, nil
+}
+
+func (p *GoTLSProgram) ConfigureManager(m *manager.Manager) {
+	// TODO check if we support go TLS on the current arch
+	if p == nil {
+		return
+	}
+
+	p.manager = m
+	p.manager.Maps = append(p.manager.Maps, []*manager.Map{
+		{Name: probeDataMap},
+		{Name: readPartialCallsMap},
+	}...)
+	// Hooks will be added in runtime for each binary
+}
+
+func (p *GoTLSProgram) ConfigureOptions(options *manager.Options) {}
+
+func (p *GoTLSProgram) Start() {
+	if p == nil {
+		return
+	}
+	// In the future Start() should just initiate the new processes listener
+	// and this implementation should be done for each new process found.
+
+	binPath := os.Getenv("GO_TLS_TEST")
+	f, err := os.Open(binPath)
+	if err != nil {
+		log.Errorf("could not open file %q due to %w", binPath, err)
+		return
+	}
+	defer f.Close()
+	elfFile, err := elf.NewFile(f)
+	if err != nil {
+		log.Errorf("file %q could not be parsed as elf: %w", binPath, err)
+		return
+	}
+	result, err := bininspect.InspectNewProcessBinary(elfFile, functionsConfig, structFieldsLookupFunctions)
+
+	// result and bin path are being passed as parameters as a preparation for the future when we will have a process
+	// watcher, so we will run on more than one binary in one goTLSProgram.
+	p.addInspectionResultToMap(result, binPath)
+
+	p.attachHooks(result, binPath)
```

Comment:
hmm, wouldn't it already be read & deleted by the probes?

Reply:
but who will call the deletion of the probes?

Reply:
I meant, the caller (a.k.a that function) should decide how to act if `addInspectionResultToMap` or `attachHooks` have failed
currently the internal functions decides how to operate and I thin that's wrong
because either we have the entire suite or nothing

and if `attachHooks` failed but we didn't clean prior resources it is a waste